File: backend/views/employee_routes.py
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, validator
from datetime import datetime

from controllers.employee_controller import EmployeeController
from controllers.settings_controller import SettingsController
from models.user import User
from dependencies import get_current_user

router = APIRouter(prefix="/api/employees", tags=["employees"])

# Pydantic models for request/response
class EmployeeCreate(BaseModel):
    full_name: str
    department: str
    position: str
    line_manager_id: int

class EmployeeUpdate(BaseModel):
    full_name: Optional[str] = None
    department: Optional[str] = None
    position: Optional[str] = None

class EmployeeResponse(BaseModel):
    id: int
    full_name: str
    department: str
    position: str
    line_manager_id: int
    status: str

class ScheduleUpdate(BaseModel):
    month: int
    year: int
    reserved_hours_per_day: float

    @validator('month')
    def validate_month(cls, v):
        if v < 1 or v > 12:
            raise ValueError('Month must be between 1 and 12')
        return v

    @validator('reserved_hours_per_day')
    def validate_hours(cls, v):
        # Basic range check - more specific validation done in route handler
        if v < 0 or v > 24:
            raise ValueError('Reserved hours must be between 0 and 24')
        return v

class EmployeeBusinessRulesUpdate(BaseModel):
    work_hours_per_day: Optional[float] = None
    work_days_per_month: Optional[float] = None
    months_in_year: Optional[int] = None

@router.post("", response_model=Dict[str, Any])
async def create_employee(employee: EmployeeCreate, current_user: User = Depends(get_current_user)):
    """Create a new employee"""
    employee_data = employee.dict()
    
    # Line managers can only create employees under themselves
    if current_user.role == 'line_manager':
        employee_data['line_manager_id'] = current_user.id
    
    result = EmployeeController.create_employee(employee_data)
    if 'error' in result:
        raise HTTPException(status_code=400, detail=result['error'])
    return result

@router.get("", response_model=List[EmployeeResponse])
async def get_employees(
    current_user: User = Depends(get_current_user),
    manager_id: Optional[int] = Query(None, description="Filter by line manager"),
    department: Optional[str] = Query(None, description="Filter by department")
):
    """Get list of employees"""
    from models.employee import Employee
    from models.user import User
    
    # Line managers can only see their own employees
    if current_user.role == 'line_manager':
        manager_id = current_user.id
    
    if manager_id:
        employees = Employee.get_by_line_manager(manager_id)
    else:
        employees = Employee.get_all_active()
    
    if department:
        employees = [e for e in employees if e.department == department]
    
    return [e.to_dict() for e in employees]

@router.get("/{employee_id}", response_model=Dict[str, Any])
async def get_employee(employee_id: int, current_user: User = Depends(get_current_user)):
    """Get employee details"""
    from models.employee import Employee
    
    # Get employee first to check line manager
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Line managers can only see their own employees
    if current_user.role == 'line_manager' and employee.line_manager_id != current_user.id:
        raise HTTPException(status_code=403, detail="Access denied: You can only view your own employees")
    
    result = EmployeeController.get_employee_by_id(employee_id)
    return result

@router.put("/{employee_id}", response_model=Dict[str, Any])
async def update_employee(employee_id: int, employee_data: EmployeeUpdate, current_user: User = Depends(get_current_user)):
    """Update employee details"""
    from models.employee import Employee
    
    # Get employee first to check line manager
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Line managers can only update their own employees
    if current_user.role == 'line_manager' and employee.line_manager_id != current_user.id:
        raise HTTPException(status_code=403, detail="Access denied: You can only update your own employees")
    
    # Convert to dict and remove None values
    update_data = {k: v for k, v in employee_data.dict().items() if v is not None}
    
    result = EmployeeController.update_employee(employee_id, update_data)
    if 'error' in result:
        raise HTTPException(status_code=400, detail=result['error'])
    return result

@router.post("/{employee_id}/update", response_model=Dict[str, Any])
async def update_employee_post(employee_id: int, employee_data: EmployeeUpdate, current_user: User = Depends(get_current_user)):
    """Update employee details (POST workaround for PUT blocking)"""
    from models.employee import Employee
    
    # Get employee first to check line manager
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Line managers can only update their own employees
    if current_user.role == 'line_manager' and employee.line_manager_id != current_user.id:
        raise HTTPException(status_code=403, detail="Access denied: You can only update your own employees")
    
    # Convert to dict and remove None values
    update_data = {k: v for k, v in employee_data.dict().items() if v is not None}
    
    result = EmployeeController.update_employee(employee_id, update_data)
    if 'error' in result:
        raise HTTPException(status_code=400, detail=result['error'])
    return result

@router.put("/{employee_id}/schedule", response_model=Dict[str, Any])
async def update_employee_schedule(employee_id: int, schedule: ScheduleUpdate, current_user: User = Depends(get_current_user)):
    """Update employee schedule"""
    from models.employee import Employee
    
    # Get employee first to check line manager
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Line managers can only update their own employees
    if current_user.role == 'line_manager' and employee.line_manager_id != current_user.id:
        raise HTTPException(status_code=403, detail="Access denied: You can only update your own employees")
    
    result = EmployeeController.update_employee_schedule(
        employee_id, 
        schedule.month, 
        schedule.year, 
        schedule.reserved_hours_per_day
    )
    if 'error' in result:
        raise HTTPException(status_code=400, detail=result['error'])
    return result

@router.get("/{employee_id}/schedule", response_model=Dict[str, Any])
async def get_employee_schedule(
    employee_id: int,
    current_user: User = Depends(get_current_user),
    year: Optional[int] = Query(None, description="Year for schedule")
):
    """Get employee schedule details"""
    from models.employee import Employee
    
    # Get employee first to check line manager
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Line managers can only see their own employees' schedules
    if current_user.role == 'line_manager' and employee.line_manager_id != current_user.id:
        raise HTTPException(status_code=403, detail="Access denied: You can only view your own employees' schedules")
    
    result = EmployeeController.get_employee_schedule_details(employee_id, year)
    if 'error' in result:
        raise HTTPException(status_code=404, detail=result['error'])
    return result

@router.get("/{employee_id}/availability/{month}/{year}", response_model=Dict[str, Any])
async def get_employee_availability(employee_id: int, month: int, year: int):
    """Check employee availability for a specific month"""
    from models.schedule import EmployeeSchedule
    from models.employee import Employee
    
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    schedule = EmployeeSchedule.get_employee_schedule(employee_id, month, year)
    if not schedule:
        raise HTTPException(status_code=404, detail="Schedule not found for this period")
    
    # Get already booked hours - sum bookings that overlap with this month
    from database import db
    query = '''
        SELECT SUM(booked_hours) as booked_hours
        FROM project_bookings
        WHERE employee_id = ? 
          AND status != 'cancelled'
          AND strftime('%Y', start_date) <= CAST(? AS TEXT)
          AND strftime('%Y', end_date) >= CAST(? AS TEXT)
          AND CAST(strftime('%m', start_date) AS INTEGER) <= ?
          AND CAST(strftime('%m', end_date) AS INTEGER) >= ?
    '''
    result = db.fetch_one(query, (employee_id, year, year, month, month))
    booked_hours = result['booked_hours'] or 0
    
    # Include reserved hours as part of the booked hours for utilization
    work_days_per_month = SettingsController.get_settings_for_employee(employee_id)['work_days_per_month']
    reserved_hours_monthly = schedule.reserved_hours_per_day * work_days_per_month
    total_utilized_hours = booked_hours + reserved_hours_monthly
    
    available_hours = schedule.get_available_hours() - booked_hours
    
    return {
        'employee': employee.to_dict(),
        'schedule': schedule.to_dict(),
        'availability': {
            'total_available': schedule.get_available_hours(),
            'already_booked': booked_hours,
            'reserved_hours': reserved_hours_monthly,
            'total_utilized': total_utilized_hours,
            'currently_available': max(0, available_hours)
        }
    }

@router.get("/{employee_id}/projects/{month}/{year}", response_model=List[Dict[str, Any]])
async def get_employee_projects(employee_id: int, month: int, year: int):
    """Get employee's project bookings for a specific month"""
    from models.employee import Employee
    
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Get project bookings for the employee that overlap with this month
    from database import db
    query = '''
        SELECT pb.*, p.name as project_name, p.project_code, p.status as project_status,
               p.attachments
        FROM project_bookings pb
        JOIN projects p ON pb.project_id = p.id
        WHERE pb.employee_id = ? 
          AND strftime('%Y', pb.start_date) <= CAST(? AS TEXT)
          AND strftime('%Y', pb.end_date) >= CAST(? AS TEXT)
          AND CAST(strftime('%m', pb.start_date) AS INTEGER) <= ?
          AND CAST(strftime('%m', pb.end_date) AS INTEGER) >= ?
        ORDER BY pb.start_date DESC
    '''
    bookings = db.fetch_all(query, (employee_id, year, year, month, month))
    
    # Parse attachments JSON for each booking
    import json
    for booking in bookings:
        if booking.get('attachments'):
            try:
                booking['attachments'] = json.loads(booking['attachments'])
            except:
                booking['attachments'] = []
        else:
            booking['attachments'] = []
    
    return bookings

@router.get("/{employee_id}/availability-range", response_model=Dict[str, Any])
async def get_employee_availability_for_date_range(
    employee_id: int,
    start_date: str = Query(..., description="Start date in YYYY-MM-DD format"),
    end_date: str = Query(..., description="End date in YYYY-MM-DD format")
):
    """Get employee's bookings, reservations, and available hours for a specific date range"""
    from models.employee import Employee
    from models.reservation import EmployeeReservation
    from database import db
    from datetime import datetime, timedelta
    
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    # Validate dates
    try:
        start = datetime.strptime(start_date, '%Y-%m-%d').date()
        end = datetime.strptime(end_date, '%Y-%m-%d').date()
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid date format. Use YYYY-MM-DD")
    
    if end < start:
        raise HTTPException(status_code=400, detail="End date must be after start date")
    
    # Get existing bookings that overlap with this date range
    bookings_query = '''
        SELECT pb.*, p.name as project_name, p.project_code
        FROM project_bookings pb
        JOIN projects p ON pb.project_id = p.id
        WHERE pb.employee_id = ? 
          AND pb.status != 'cancelled'
          AND (
              (pb.start_date <= ? AND pb.end_date >= ?)
              OR (pb.start_date <= ? AND pb.end_date >= ?)
              OR (pb.start_date >= ? AND pb.end_date <= ?)
          )
        ORDER BY pb.start_date
    '''
    bookings = db.fetch_all(bookings_query, (
        employee_id,
        start, start,
        end, end,
        start, end
    ))
    
    # Get reservations that overlap with this date range
    reservations = EmployeeReservation.get_active_reservations_for_date_range(
        employee_id, start, end
    )
    
    # Calculate working days in the requested range
    def count_working_days(start_date, end_date):
        working_days = 0
        current = start_date
        while current <= end_date:
            # Weekend = Friday(4) + Saturday(5)
            if current.weekday() not in (4, 5):
                working_days += 1
            current += timedelta(days=1)
        return working_days
    
    # Calculate available hours considering overlapping bookings and reservations
    # We need to track utilization per day
    total_working_days = count_working_days(start, end)
    max_hours_per_day = SettingsController.get_settings_for_employee(employee_id)['work_hours_per_day']
    total_max_hours = total_working_days * max_hours_per_day
    
    # Calculate total booked hours for overlapping bookings
    # For bookings, we need to calculate how many hours fall within our date range
    total_booked_hours = 0
    for booking in bookings:
        b_start = booking['start_date'] if isinstance(booking['start_date'], type(start)) else datetime.strptime(str(booking['start_date']), '%Y-%m-%d').date()
        b_end = booking['end_date'] if isinstance(booking['end_date'], type(end)) else datetime.strptime(str(booking['end_date']), '%Y-%m-%d').date()
        
        # Find overlap between booking and requested range
        overlap_start = max(start, b_start)
        overlap_end = min(end, b_end)
        
        if overlap_start <= overlap_end:
            # Calculate working days in the overlap
            overlap_working_days = count_working_days(overlap_start, overlap_end)
            booking_total_working_days = count_working_days(b_start, b_end)
            
            # Pro-rate the booked hours based on the overlap
            if booking_total_working_days > 0:
                hours_per_day = booking['booked_hours'] / booking_total_working_days
                overlap_hours = hours_per_day * overlap_working_days
                total_booked_hours += overlap_hours
    
    # Calculate reserved hours
    # For reservations, hours are per day, so we count days in overlap and multiply
    total_reserved_hours = 0
    for reservation in reservations:
        r_start = reservation.start_date if isinstance(reservation.start_date, type(start)) else datetime.strptime(str(reservation.start_date), '%Y-%m-%d').date()
        r_end = reservation.end_date if isinstance(reservation.end_date, type(end)) else datetime.strptime(str(reservation.end_date), '%Y-%m-%d').date()
        
        # Find overlap between reservation and requested range
        overlap_start = max(start, r_start)
        overlap_end = min(end, r_end)
        
        if overlap_start <= overlap_end:
            overlap_working_days = count_working_days(overlap_start, overlap_end)
            total_reserved_hours += reservation.reserved_hours_per_day * overlap_working_days
    
    total_utilized_hours = total_booked_hours + total_reserved_hours
    available_hours = max(0, total_max_hours - total_utilized_hours)
    
    # Calculate max bookable hours per day on average
    if total_working_days > 0:
        avg_utilized_per_day = total_utilized_hours / total_working_days
        avg_available_per_day = max(0, max_hours_per_day - avg_utilized_per_day)
    else:
        avg_utilized_per_day = 0
        avg_available_per_day = max_hours_per_day
    
    return {
        'employee': employee.to_dict(),
        'date_range': {
            'start_date': start_date,
            'end_date': end_date
        },
        'bookings': [dict(b) for b in bookings],
        'reservations': [r.to_dict() for r in reservations],
        'availability': {
            'working_days': total_working_days,
            'max_hours_total': total_max_hours,
            'total_booked_hours': round(total_booked_hours, 1),
            'total_reserved_hours': round(total_reserved_hours, 1),
            'total_utilized_hours': round(total_utilized_hours, 1),
            'available_hours': round(available_hours, 1),
            'avg_utilized_per_day': round(avg_utilized_per_day, 2),
            'avg_available_per_day': round(avg_available_per_day, 2)
        }
    }

@router.get("/{employee_id}/business-rules", response_model=Dict[str, Any])
async def get_employee_business_rules(employee_id: int, current_user: User = Depends(get_current_user)):
    """Get custom business rules for an employee. Returns null fields if using global defaults."""
    from models.employee import Employee
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    rules = SettingsController.get_employee_business_rules(employee_id)
    effective = SettingsController.get_settings_for_employee(employee_id)
    return {
        'employee_id': employee_id,
        'custom_rules': rules,
        'effective_settings': effective
    }

@router.put("/{employee_id}/business-rules", response_model=Dict[str, Any])
async def set_employee_business_rules(
    employee_id: int,
    rules: EmployeeBusinessRulesUpdate,
    current_user: User = Depends(get_current_user)
):
    """Set or update custom business rules for an employee."""
    from models.employee import Employee
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    update_data = {k: v for k, v in rules.dict().items() if v is not None}
    if not update_data:
        raise HTTPException(status_code=400, detail="No fields provided to update")
    saved = SettingsController.set_employee_business_rules(employee_id, update_data)
    effective = SettingsController.get_settings_for_employee(employee_id)
    return {
        'employee_id': employee_id,
        'custom_rules': saved,
        'effective_settings': effective
    }

@router.delete("/{employee_id}/business-rules", response_model=Dict[str, Any])
async def delete_employee_business_rules(employee_id: int, current_user: User = Depends(get_current_user)):
    """Remove custom business rules for an employee, reverting to global settings."""
    from models.employee import Employee
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    SettingsController.delete_employee_business_rules(employee_id)
    return {
        'employee_id': employee_id,
        'message': 'Custom business rules removed. Employee will now use global settings.',
        'effective_settings': SettingsController.get_settings()
    }

@router.delete("/{employee_id}", response_model=Dict[str, Any])
async def delete_employee(employee_id: int):
    """Delete an employee"""
    result = EmployeeController.delete_employee(employee_id)
    if 'error' in result:
        status_code = 404 if result['error'] == 'Employee not found' else 400
        raise HTTPException(status_code=status_code, detail=result['error'])
    return result

File: backend-node/src/routes/employeeRoutes.js
const express = require('express');
const router = express.Router();
const EmployeeController = require('../controllers/employeeController');
const SettingsController = require('../controllers/settingsController');
const Employee = require('../models/employee');
const EmployeeSchedule = require('../models/schedule');
const EmployeeReservation = require('../models/reservation');
const { authenticate } = require('../middleware/auth');
const db = require('../database');

// Create employee
router.post('/', authenticate, (req, res) => {
  try {
    let employeeData = req.body;

    // Line managers can only create employees under themselves
    if (req.currentUser.role === 'line_manager') {
      employeeData.line_manager_id = req.currentUser.id;
    }

    const result = EmployeeController.createEmployee(employeeData);
    if (result.error) {
      return res.status(400).json({ detail: result.error });
    }
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get all employees
router.get('/', authenticate, (req, res) => {
  try {
    let managerId = req.query.manager_id ? parseInt(req.query.manager_id) : null;
    const department = req.query.department;

    // Line managers can only see their own employees
    if (req.currentUser.role === 'line_manager') {
      managerId = req.currentUser.id;
    }

    let employees;
    if (managerId) {
      employees = Employee.getByLineManager(managerId);
    } else {
      employees = Employee.getAllActive();
    }

    if (department) {
      employees = employees.filter(e => e.department === department);
    }

    res.json(employees.map(e => e.toDict()));
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get employee by ID
router.get('/:employeeId', authenticate, (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const employee = Employee.getById(employeeId);

    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    // Line managers can only see their own employees
    if (req.currentUser.role === 'line_manager' && employee.line_manager_id !== req.currentUser.id) {
      return res.status(403).json({ detail: 'Access denied: You can only view your own employees' });
    }

    const result = EmployeeController.getEmployeeById(employeeId);
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Update employee (PUT)
router.put('/:employeeId', authenticate, (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const employee = Employee.getById(employeeId);

    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    // Line managers can only update their own employees
    if (req.currentUser.role === 'line_manager' && employee.line_manager_id !== req.currentUser.id) {
      return res.status(403).json({ detail: 'Access denied: You can only update your own employees' });
    }

    const result = EmployeeController.updateEmployee(employeeId, req.body);
    if (result.error) {
      return res.status(400).json({ detail: result.error });
    }
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Update employee (POST workaround)
router.post('/:employeeId/update', authenticate, (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const employee = Employee.getById(employeeId);

    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    // Line managers can only update their own employees
    if (req.currentUser.role === 'line_manager' && employee.line_manager_id !== req.currentUser.id) {
      return res.status(403).json({ detail: 'Access denied: You can only update your own employees' });
    }

    const result = EmployeeController.updateEmployee(employeeId, req.body);
    if (result.error) {
      return res.status(400).json({ detail: result.error });
    }
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Update employee schedule
router.put('/:employeeId/schedule', authenticate, (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const employee = Employee.getById(employeeId);

    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    // Line managers can only update their own employees
    if (req.currentUser.role === 'line_manager' && employee.line_manager_id !== req.currentUser.id) {
      return res.status(403).json({ detail: 'Access denied: You can only update your own employees' });
    }

    const { month, year, reserved_hours_per_day } = req.body;

    if (month < 1 || month > 12) {
      return res.status(400).json({ detail: 'Month must be between 1 and 12' });
    }

    if (reserved_hours_per_day < 0 || reserved_hours_per_day > 24) {
      return res.status(400).json({ detail: 'Reserved hours must be between 0 and 24' });
    }

    const result = EmployeeController.updateEmployeeSchedule(employeeId, month, year, reserved_hours_per_day);
    if (result.error) {
      return res.status(400).json({ detail: result.error });
    }
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get employee schedule
router.get('/:employeeId/schedule', authenticate, (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const year = req.query.year ? parseInt(req.query.year) : null;
    const employee = Employee.getById(employeeId);

    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    // Line managers can only see their own employees' schedules
    if (req.currentUser.role === 'line_manager' && employee.line_manager_id !== req.currentUser.id) {
      return res.status(403).json({ detail: "Access denied: You can only view your own employees' schedules" });
    }

    const result = EmployeeController.getEmployeeScheduleDetails(employeeId, year);
    if (result.error) {
      return res.status(404).json({ detail: result.error });
    }
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get employee availability for a specific month
router.get('/:employeeId/availability/:month/:year', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const month = parseInt(req.params.month);
    const year = parseInt(req.params.year);

    const employee = Employee.getById(employeeId);
    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    const schedule = EmployeeSchedule.getEmployeeSchedule(employeeId, month, year);
    if (!schedule) {
      return res.status(404).json({ detail: 'Schedule not found for this period' });
    }

    // Get already booked hours
    const result = db.fetchOne(`
      SELECT SUM(booked_hours) as booked_hours
      FROM project_bookings
      WHERE employee_id = ?
        AND status != 'cancelled'
        AND strftime('%Y', start_date) <= CAST(? AS TEXT)
        AND strftime('%Y', end_date) >= CAST(? AS TEXT)
        AND CAST(strftime('%m', start_date) AS INTEGER) <= ?
        AND CAST(strftime('%m', end_date) AS INTEGER) >= ?
    `, [employeeId, year, year, month, month]);

    const bookedHours = result.booked_hours || 0;

    // Include reserved hours as part of the booked hours for utilization
    const workDaysPerMonth = SettingsController.getWorkDaysPerMonth();
    const reservedHoursMonthly = schedule.reserved_hours_per_day * workDaysPerMonth;
    const totalUtilizedHours = bookedHours + reservedHoursMonthly;

    const availableHours = schedule.getAvailableHours() - bookedHours;

    res.json({
      employee: employee.toDict(),
      schedule: schedule.toDict(),
      availability: {
        total_available: schedule.getAvailableHours(),
        already_booked: bookedHours,
        reserved_hours: reservedHoursMonthly,
        total_utilized: totalUtilizedHours,
        currently_available: Math.max(0, availableHours)
      }
    });
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get employee's projects for a specific month
router.get('/:employeeId/projects/:month/:year', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const month = parseInt(req.params.month);
    const year = parseInt(req.params.year);

    const employee = Employee.getById(employeeId);
    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    const bookings = db.fetchAll(`
      SELECT pb.*, p.name as project_name, p.project_code, p.status as project_status,
             p.attachments
      FROM project_bookings pb
      JOIN projects p ON pb.project_id = p.id
      WHERE pb.employee_id = ?
        AND strftime('%Y', pb.start_date) <= CAST(? AS TEXT)
        AND strftime('%Y', pb.end_date) >= CAST(? AS TEXT)
        AND CAST(strftime('%m', pb.start_date) AS INTEGER) <= ?
        AND CAST(strftime('%m', pb.end_date) AS INTEGER) >= ?
      ORDER BY pb.start_date DESC
    `, [employeeId, year, year, month, month]);

    // Parse attachments JSON for each booking
    const result = bookings.map(booking => {
      const b = { ...booking };
      if (b.attachments) {
        try {
          b.attachments = JSON.parse(b.attachments);
        } catch {
          b.attachments = [];
        }
      } else {
        b.attachments = [];
      }
      return b;
    });

    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get employee availability for a date range
router.get('/:employeeId/availability-range', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const startDateStr = req.query.start_date;
    const endDateStr = req.query.end_date;

    if (!startDateStr || !endDateStr) {
      return res.status(400).json({ detail: 'start_date and end_date are required' });
    }

    const employee = Employee.getById(employeeId);
    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    // Parse and validate dates
    const start = new Date(startDateStr);
    const end = new Date(endDateStr);

    if (isNaN(start.getTime()) || isNaN(end.getTime())) {
      return res.status(400).json({ detail: 'Invalid date format. Use YYYY-MM-DD' });
    }

    if (end < start) {
      return res.status(400).json({ detail: 'End date must be after start date' });
    }

    // Get existing bookings that overlap with this date range
    const bookings = db.fetchAll(`
      SELECT pb.*, p.name as project_name, p.project_code
      FROM project_bookings pb
      JOIN projects p ON pb.project_id = p.id
      WHERE pb.employee_id = ?
        AND pb.status != 'cancelled'
        AND (
          (pb.start_date <= ? AND pb.end_date >= ?)
          OR (pb.start_date <= ? AND pb.end_date >= ?)
          OR (pb.start_date >= ? AND pb.end_date <= ?)
        )
      ORDER BY pb.start_date
    `, [
      employeeId,
      startDateStr, startDateStr,
      endDateStr, endDateStr,
      startDateStr, endDateStr
    ]);

    // Get reservations that overlap with this date range
    const reservations = EmployeeReservation.getActiveReservationsForDateRange(employeeId, startDateStr, endDateStr);

    // Calculate working days
    function countWorkingDays(startDate, endDate) {
      let workingDays = 0;
      const current = new Date(startDate);
      const endD = new Date(endDate);

      while (current <= endD) {
        if (current.getDay() !== 0 && current.getDay() !== 6) {
          workingDays++;
        }
        current.setDate(current.getDate() + 1);
      }
      return workingDays;
    }

    const totalWorkingDays = countWorkingDays(start, end);
    const maxHoursPerDay = SettingsController.getWorkHoursPerDay();
    const totalMaxHours = totalWorkingDays * maxHoursPerDay;

    // Calculate total booked hours for overlapping bookings
    let totalBookedHours = 0;
    for (const booking of bookings) {
      const bStart = new Date(booking.start_date);
      const bEnd = new Date(booking.end_date);

      const overlapStart = new Date(Math.max(start.getTime(), bStart.getTime()));
      const overlapEnd = new Date(Math.min(end.getTime(), bEnd.getTime()));

      if (overlapStart <= overlapEnd) {
        const overlapWorkingDays = countWorkingDays(overlapStart, overlapEnd);
        const bookingTotalWorkingDays = countWorkingDays(bStart, bEnd);

        if (bookingTotalWorkingDays > 0) {
          const hoursPerDay = booking.booked_hours / bookingTotalWorkingDays;
          const overlapHours = hoursPerDay * overlapWorkingDays;
          totalBookedHours += overlapHours;
        }
      }
    }

    // Calculate reserved hours
    let totalReservedHours = 0;
    for (const reservation of reservations) {
      const rStart = new Date(reservation.start_date);
      const rEnd = new Date(reservation.end_date);

      const overlapStart = new Date(Math.max(start.getTime(), rStart.getTime()));
      const overlapEnd = new Date(Math.min(end.getTime(), rEnd.getTime()));

      if (overlapStart <= overlapEnd) {
        const overlapWorkingDays = countWorkingDays(overlapStart, overlapEnd);
        totalReservedHours += reservation.reserved_hours_per_day * overlapWorkingDays;
      }
    }

    const totalUtilizedHours = totalBookedHours + totalReservedHours;
    const availableHours = Math.max(0, totalMaxHours - totalUtilizedHours);

    let avgUtilizedPerDay = 0;
    let avgAvailablePerDay = maxHoursPerDay;
    if (totalWorkingDays > 0) {
      avgUtilizedPerDay = totalUtilizedHours / totalWorkingDays;
      avgAvailablePerDay = Math.max(0, maxHoursPerDay - avgUtilizedPerDay);
    }

    res.json({
      employee: employee.toDict(),
      date_range: {
        start_date: startDateStr,
        end_date: endDateStr
      },
      bookings,
      reservations: reservations.map(r => r.toDict()),
      availability: {
        working_days: totalWorkingDays,
        max_hours_total: totalMaxHours,
        total_booked_hours: Math.round(totalBookedHours * 10) / 10,
        total_reserved_hours: Math.round(totalReservedHours * 10) / 10,
        total_utilized_hours: Math.round(totalUtilizedHours * 10) / 10,
        available_hours: Math.round(availableHours * 10) / 10,
        avg_utilized_per_day: Math.round(avgUtilizedPerDay * 100) / 100,
        avg_available_per_day: Math.round(avgAvailablePerDay * 100) / 100
      }
    });
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Delete employee
router.delete('/:employeeId', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const result = EmployeeController.deleteEmployee(employeeId);

    if (result.error) {
      const statusCode = result.error === 'Employee not found' ? 404 : 400;
      return res.status(statusCode).json({ detail: result.error });
    }
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

module.exports = router;
